File: python/systemdesign/url_shortener.py
"""
URL shortener that encodes long URLs into short base62 codes and
decodes them back, staying collision-safe by keying off an
incrementing counter rather than hashing the URL (so two encodes of
the same URL round-trip through the same stored mapping).

status - completed
"""

from base_logger.logging_event import create_logger

ALPHABET = "0123456789ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz"
BASE = len(ALPHABET)


class UrlShortener:

    log = create_logger(__name__)

    def __init__(self, base_url="https://short.ly/"):
        self.base_url = base_url
        self.url_to_code = {}
        self.code_to_url = {}
        self._next_id = 1

    def _encode_id(self, num: int) -> str:
        if num == 0:
            return ALPHABET[0]
        digits = []
        while num > 0:
            num, remainder = divmod(num, BASE)
            digits.append(ALPHABET[remainder])
        return "".join(reversed(digits))

    def shorten(self, long_url: str) -> str:
        if long_url in self.url_to_code:
            code = self.url_to_code[long_url]
        else:
            code = self._encode_id(self._next_id)
            self._next_id += 1
            self.url_to_code[long_url] = code
            self.code_to_url[code] = long_url
        return self.base_url + code

    def expand(self, short_url: str) -> str:
        code = short_url[len(self.base_url):]
        if code not in self.code_to_url:
            raise KeyError(f"unknown short code: {code}")
        return self.code_to_url[code]


if __name__ == "__main__":
    log = create_logger("url_shortener_demo")

    shortener = UrlShortener()

    short1 = shortener.shorten("https://example.com/very/long/path/about/system/design")
    short2 = shortener.shorten("https://example.com/another/long/article")
    log.info("shortened url 1 -> %s", short1)
    log.info("shortened url 2 -> %s", short2)
    log.info("codes are distinct: %s", short1 != short2)

    short1_again = shortener.shorten("https://example.com/very/long/path/about/system/design")
    log.info("re-shortening the same url -> %s (expected same code as before)", short1_again)
    log.info("collision-safe (no new code minted): %s", short1 == short1_again)

    expanded = shortener.expand(short1)
    log.info("expand(%s) -> %s", short1, expanded)

    try:
        shortener.expand("https://short.ly/doesnotexist")
    except KeyError as exc:
        log.info("expanding unknown code raised: %s", exc)

File: typescript/systemdesign/urlShortener.ts
/*
URL shortener: encodes long URLs into short base62 codes and decodes them
back, with collision-safe storage (a counter-based code plus a reverse map
so the same long URL always maps back to its one short code).

status - completed
*/

import { logp } from "../utils/logger";

const BASE62_ALPHABET = "0123456789ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz";

function encodeBase62(num: number): string {
  if (num === 0) return BASE62_ALPHABET[0];
  let result = "";
  let n = num;
  while (n > 0) {
    result = BASE62_ALPHABET[n % 62] + result;
    n = Math.floor(n / 62);
  }
  return result;
}

class UrlShortener {
  private readonly longToCode: Map<string, string> = new Map();
  private readonly codeToLong: Map<string, string> = new Map();
  private nextId = 1;

  constructor(private readonly baseUrl: string = "https://short.ly/") {}

  shorten(longUrl: string): string {
    const existingCode = this.longToCode.get(longUrl);
    if (existingCode) {
      return this.baseUrl + existingCode;
    }

    let code: string;
    do {
      code = encodeBase62(this.nextId++);
    } while (this.codeToLong.has(code)); // guards against any future non-monotonic id source

    this.longToCode.set(longUrl, code);
    this.codeToLong.set(code, longUrl);
    return this.baseUrl + code;
  }

  expand(shortUrl: string): string | undefined {
    const code = shortUrl.startsWith(this.baseUrl) ? shortUrl.slice(this.baseUrl.length) : shortUrl;
    return this.codeToLong.get(code);
  }
}

// usage scenarios

const shortener = new UrlShortener();

const short1 = shortener.shorten("https://example.com/some/very/long/path?query=1");
const short2 = shortener.shorten("https://example.com/another/long/path");
logp(`shortened url 1 -> ${short1}`);
logp(`shortened url 2 -> ${short2}`);

const short1Again = shortener.shorten("https://example.com/some/very/long/path?query=1");
logp(`shortening the same long url again returns the same code -> ${short1Again === short1}`);

logp(`expand(short1) -> ${shortener.expand(short1)}`);
logp(`expand(short2) -> ${shortener.expand(short2)}`);
logp(`expand of unknown code -> ${shortener.expand("https://short.ly/ZZZZZ")} (expected undefined)`);

const many: string[] = [];
for (let i = 0; i < 5; i++) {
  many.push(shortener.shorten(`https://example.com/page-${i}`));
}
const allUnique = new Set(many).size === many.length;
logp(`5 distinct urls produced ${many.length} codes, all unique: ${allUnique} -> ${many.join(", ")}`);
